communication-refactored/src/interface/swarm_task.rs

```diff
@@ -0,0 +1,571 @@
+    // Response channels for start-listening via a relay.
+    // A result is returned once the associated listener reported it's first new listening address, or a listener error
+    // occurred. Additionally, an error will be returned if the relay could not be connected.
+    #[cfg(feature = "relay")]
+    await_relayed_listen: HashMap<ListenerId, (PeerId, oneshot::Sender<Result<Multiaddr, ListenRelayErr>>)>,
+}
+
+impl<Rq, Rs, TRq> SwarmTask<Rq, Rs, TRq>
+where
+    Rq: RqRsMessage + Borrow<TRq>,
+    Rs: RqRsMessage,
+    TRq: Clone + Send + 'static,
+{
+    // Create new instance of a swarm-task.
+    pub fn new(
+        swarm: Swarm<NetBehaviour<Rq, Rs, TRq>>,
+        command_rx: mpsc::Receiver<SwarmOperation<Rq, Rs, TRq>>,
+        request_tx: mpsc::Sender<ReceiveRequest<Rq, Rs>>,
+        event_tx: Option<mpsc::Sender<NetworkEvent>>,
+    ) -> Self {
+        SwarmTask {
+            swarm,
+            command_rx,
+            request_tx,
+            event_tx,
+            listeners: HashMap::new(),
+            await_response: HashMap::new(),
+            await_connection: HashMap::new(),
+            await_listen: HashMap::new(),
+            #[cfg(feature = "relay")]
+            await_relayed_listen: HashMap::new(),
+        }
+    }
+
+    // Central loop:
+    // - Drive the [`Swarm`] by polling it for events.
+    // - Poll the commands-channel for [`SwarmOperation`]s that are sent from [`ShCommunication`].
+    //
+    // If all [`ShCommunication`] clones are dropped, the command-channel will return `None` and [`SwarmTask`] will shut
+    // down.
+    pub async fn run(mut self) {
+        loop {
+            futures::select! {
+                // Receive [`SwarmOperation`]s to initiate operations on the [`Swarm`].
+                command = self.command_rx.next().fuse() => {
+                    if let Some(c) = command {
+                        self.handle_command(c)
+                    } else {
+                        break;
+                    }
+                },
+                // Drive the swarm and handle events
+                event = self.swarm.select_next_some() => self.handle_swarm_event(event)
+            }
+        }
+        self.shutdown();
+    }
+
+    // Check if the swarm events yields a result for a previously initiated operation.
+    // Optionally forward a [`NetworkEvent`] for the event.
+    fn handle_swarm_event<THandleErr>(&mut self, event: SwarmEvent<BehaviourEvent<Rq, Rs>, THandleErr>) {
+        match event {
+            SwarmEvent::Behaviour(BehaviourEvent::ReceivedRequest {
+                request_id,
+                peer,
+                request,
+                response_tx,
+            }) => {
+                let received_rq = ReceiveRequest {
+                    request_id,
+                    peer,
+                    request,
+                    response_tx,
+                };
+                let _ = self.request_tx.try_send(received_rq);
```

Comment:
Is sending these to the _inbound request handler_ on a best-effort basis good enough? Why not properly send them via `SinkExt::send`?

Reply:
But wouldn't `SinktExt::send` require to `await` the returned future, hence potentially block everything? 
And I am not sure how this would behave in case that the `Receiver` is not dropped, but also never read. 

I have considered adding a RingBuffer structure to cache requests in case that `try_send` returns `SendErrorKind::Full`.  In that case the loop in `SwarmTask::run` would also always do something like
```
futures::select! {
      ...
      _ = poll_fn(|cx| request_tx.poll_ready(cx)).fuse() => {
          // send oldest element from ring buffer
      }
  }
```
The benefit from using a ring buffer would be that only the most recent requests would be cached.
Another option would be to use an unbounded channel for inbound requests, but this could potentially use a lot of memory and even be a security risk if a remote is spam-sending requests.

I so far kept it rather simple because usually I'd expect the `Receiver` to have some sort of loop that continuously reads from it (and if the `Receiver` is dropped there nothing we can do anyway). But I agree that this is probably not good enough.
What do you think about the ring-buffer idea @mxinden?

Reply:
> But wouldn't SinktExt::send require to await the returned future, hence potentially block everything?

That is correct. You would depend on the other side to continuously read from the channel. Is there reason to worry that the other side will not continuously read from that channel?

> Another option would be to use an unbounded channel for inbound requests, but this could potentially use a lot of memory and even be a security risk if a remote is spam-sending requests.

I would strongly advice against unbounded channels. See for example https://github.com/paritytech/substrate/search?q=unbounded&type=issues for the many issues we had related to unbounded channels.

>  What do you think about the ring-buffer idea @mxinden?

While I think it is a healthy attitude to drop old messages in favor of new ones, I don't quite see the need for the ring buffer in the first place, circling back to my question above, is there reason to worry that the other side will not continuously read from that channel?

Reply:
> Is there reason to worry that the other side will not continuously read from that channel?

In general, I would expect the receiver to continuously read from the channel, yes. 
But I could also imagine use-cases in which a request may require e.g. some asynchronous operations or user interaction, and where requests are handled sequentially. In this case, a next request might only be read from the channel after the preceding one was resolved, which could then potentially take some time. In the meantime, no swarm interaction could take place once the capacity of the channel was reached. I think this behaviour may be desired in some cases, but maybe not in every one.

--- 
I guess my general motive is that the `SwarmTask` should not have to rely on any behaviour outside of it's own loop. Because it is running in its own decoupled task, I'd like to guarantee that it will not block, which is why I try to not use any asynchronous operations (apart from the central `run` method). 

If the receiver continuously polls the request channel, `SinktExt::send` would not block, but in this case the current best-effort basis via `Sender::try_send` would also be sufficient, wouldn't it?

Reply:
_Take the below with a grain of salt. After all I am not very familiar with your whole stack._

> I think this behaviour may be desired in some cases, 

It would be desired to enforce back-pressure throughout the system. E.g. when the machine is already at its limit, the network should pause for a bit and not produce even more events.

> but maybe not in every one.

I would expect a large enough channel size to cater for those cases.

> And if the receiver continuously polls the request channel, SinktExt::send would not block, but in this case the current best-effort basis via Sender::try_send would also be sufficient, wouldn't it?

It would. Though (a) you loose the first mentioned back-pressure (whether that is useful in your case or not I can't tell) and (b) silently dropped events are hard to debug.

Reply:
> It would be desired to enforce back-pressure throughout the system. E.g. when the machine is already at its limit, the network should pause for a bit and not produce even more events.

That's a good point, which I did not consider before.

---
After thinking about this more, I decided to go the extra mile and support all 3 discussed configuration (block/drop/use ring-buffer). I think this behaviour should be decided by the user, especially since Stronghold-p2p is only a library and I do not know all use cases.
I have added a new commit https://github.com/iotaledger/stronghold.rs/pull/210/commits/d93c9f632bbcba336ae70e2b18b86e0e286b5111 that implements a wrapper for the `mpsc::Sender` and implements `Sink` for it according to the set config.